apm-agent-plugins/apm-scala-concurrent-plugin/src/test/scala/co/elastic/apm/agent/scala/concurrent/FutureInstrumentationSpec.scala

```diff
@@ -0,0 +1,224 @@
+      true
+    )
+  }
+
+  test("Multiple async operations should be able to set context on the current transaction") {
+
+    implicit val multiPoolEc: ExecutionContextExecutor =
+      ExecutionContext.fromExecutor(Executors.newFixedThreadPool(3))
+
+    val future = Future
+      .traverse(1 to 100) { _ =>
+        Future.sequence(List(
+          Future {
+            Thread.sleep(25)
+            tracer.currentTransaction().addCustomContext("future1", true)
+          },
+          Future {
+            Thread.sleep(50)
+            tracer.currentTransaction().addCustomContext("future2", true)
+          },
+          Future {
+            Thread.sleep(10)
+            tracer.currentTransaction().addCustomContext("future3", true)
+          }
+        ))
+      }
+
+    Await.ready(future, 10.seconds)
+    transaction.deactivate().end()
+    assertEquals(
+      reporter.getTransactions.get(0).getContext.getCustom("future1").asInstanceOf[Boolean],
+      true
+    )
+    assertEquals(
+      reporter.getTransactions.get(0).getContext.getCustom("future2").asInstanceOf[Boolean],
+      true
+    )
+    assertEquals(
+      reporter.getTransactions.get(0).getContext.getCustom("future3").asInstanceOf[Boolean],
+      true
+    )
+
+  }
+
+  test("Handle a combination of Promises and Futures correctly") {
+
+    implicit val multiPoolEc: ExecutionContextExecutor =
+      ExecutionContext.fromExecutor(Executors.newFixedThreadPool(3))
+
+    val promise = Promise[Int]()
+
+    Future {
+      Thread.sleep(100)
+      }
+      .map(_ => 42)
+      .onComplete {
+        case Success(value) => promise.success(value)
+        case Failure(exception) => promise.failure(exception)
+      }
+
+    val future = promise
+      .future
+      .map(_ => tracer.currentTransaction().addCustomContext("future", true))
+
+    Await.ready(future, 10.seconds)
+    transaction.deactivate().end()
+    assertEquals(
+      reporter.getTransactions.get(0).getContext.getCustom("future").asInstanceOf[Boolean],
+      true
+    )
+
+  }
+
+  test("Handle a combination of Promises and complex Futures correctly") {
+
+    implicit val multiPoolEc: ExecutionContextExecutor =
+      ExecutionContext.fromExecutor(Executors.newFixedThreadPool(3))
+
+    val promise = Promise[Int]()
+
+    Future
+      .traverse(1 to 100) { _ =>
+        Future.sequence(List(
+          Future {
+            Thread.sleep(25)
+          },
+          Future {
+            Thread.sleep(50)
+          },
+          Future {
+            Thread.sleep(10)
+          }
+        ))
+      }
+      .map(_ => 42)
+      .onComplete {
+        case Success(value) => promise.success(value)
+        case Failure(exception) => promise.failure(exception)
+      }
+
+    val future = promise
+      .future
+      .map(_ => tracer.currentTransaction().addCustomContext("future", true))
+
+    Await.ready(future, 10.seconds)
```

Comment:
@felixbarny I saw in your branch that you were using this approach.

Technically, this is not needed as the test framework will correctly await the Future for you.

But I can imagine that this makes it more explicit when you don't know the test-framework or Scala Futures.